presto-main/src/main/java/com/facebook/presto/operator/BlockEncodingBuffers.java

```diff
@@ -0,0 +1,442 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.ByteArrayBlock;
+import com.facebook.presto.spi.block.DictionaryBlock;
+import com.facebook.presto.spi.block.Int128ArrayBlock;
+import com.facebook.presto.spi.block.IntArrayBlock;
+import com.facebook.presto.spi.block.LongArrayBlock;
+import com.facebook.presto.spi.block.RunLengthEncodedBlock;
+import com.facebook.presto.spi.block.ShortArrayBlock;
+import io.airlift.slice.SliceOutput;
+
+import javax.annotation.Nullable;
+
+import java.util.Arrays;
+
+import static com.facebook.presto.operator.UncheckedByteArrays.fill;
+import static com.facebook.presto.operator.UncheckedByteArrays.setByte;
+import static com.google.common.base.Verify.verify;
+import static io.airlift.slice.SizeOf.SIZE_OF_BYTE;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.lang.Math.max;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+public abstract class BlockEncodingBuffers
+{
+    private static final int BITS_IN_BYTE = 8;
+    private static final int INITIAL_POSITION_COUNT = 1024;
+
+    // The initial buffer size in terms of number of elements
+    protected final int initialPositionCount;
+
+    // The block after peeling off the Dictionary or RLE wrappings.
+    protected Block decodedBlock;
+
+    // The number of positions (rows) to be copied from the incoming block.
+    protected int positionCount;
+
+    // The number of positions (rows) to be copied from the incoming block within current batch.
+    protected int batchSize;
+
+    // The offset into positions/mappedPositions array that the current batch starts copying.
+    // I.e. in each batch we copy the values of rows from positions[positionsOffset] to positions[positionsOffset + batchSize]
+    protected int positionsOffset;
+
+    // The number of positions (rows) buffered so far
+    protected int bufferedPositionCount;
+
+    // Whether the positions array was mapped to mappedPositions
+    protected boolean isPositionsMapped;
+
+    // The positions (row numbers) of the incoming Block to be copied to this buffer.
+    // All top level BlockEncodingBuffers for the same partition share the same positions array.
+    private int[] positions;
+
+    // The mapped positions of the incoming Dictionary or RLE Block
+    @Nullable
+    private int[] mappedPositions;
+
+    @Nullable
+    private byte[] nullsBuffer;
```

Comment:
I'd rename to `nulls` and `nullsIndex` for brevity and readability.

Reply:
The array/row/map/varchar have offsets array and offsetsBuffer. offsets array is a temporary holder for offsets in current block while offsetsBuffer is the offsets for all pages processed so far. How do we name them?

Reply:
I haven't reviewed that code yet. Will suggest something once I get there.

Reply:
How about renaming offsets array in Array/row/map/VariableWidthBlockEncodingBuffers to offsetsForCurrentPage and offsetsBuffer to offsets? Then we can rename all valuesBuffer to values, nullsBuffer to nulls etc. @mbasmanova @wenleix